### DISCUSSION TITLE
What is the alternate to IWorkflowInvoker? IBuildsandStartWorkflow is throwing exception with DI

### DISCUSSION THREAD

**Discussion by SachinMehndiratta** [elsa-workflows/elsa-core#958] [Q&A]
Hi,

Not sure about building and running the workflows in web app, tried with IBuildsandStartWorkflow via IServiceProvider and encountered the error of disposed object.

![image](https://user-images.githubusercontent.com/29795134/117671153-1f32dc80-b1c6-11eb-815d-b3c52dba3c5e.png)

Seems like context is disposed when saving in peristence layer and hence the error. Context must be handled via DI only.

Any guidelines or clue here?

**Comment by sfmskywalker** [comment]
Hard to say without seeing the full project (e.g. I don't see where you are setting up your DI nor do I see from where you are invoking `Tes`.

Perhaps the following quick start might help: https://elsa-workflows.github.io/elsa-core/docs/next/quickstarts/quickstarts-console-hello-world.

**Reply by SachinMehndiratta** [reply]
@sfmskywalker  - Changing the return type to Task from Void solves this, however I am a bit confused that await statement never worked and the workflow is never executed. My workflow is a CRON job and I feel this is an issue with CRON.

If I add this workflow in esla startup only then it worked fine but invoking via Controller never worked for CRON

**Reply by sfmskywalker** [reply]
Ah. So your workflow starts with the `Cron` activity, am I right? In that case, the workflow should at least execute once when I look at the `Cron`'s source code. But regardless, you shouldn't manually invoke workflows that start with Cron. Instead, you should let Elsa execute it for you (which happens automatically if you register Quartz or Hangfire temporal activities).

If, on the other hand, your workflow doesn't start with a Cron activity, but with an action-like activity such as `WriteLine`, then it should work fine. If this is the case and the workflow doesn't execute at least once and also not recurring, then something might be broken. In that case, I would love to get a repro so I can see for myself.

**Reply by SachinMehndiratta** [reply]
Yes, my workflow starts with cron activity. 

I am looking for dynamic scheduling of workflows, something like I can initiate/schedule my workflow(orchestration job) on demand and passing the CRON expression in workflow context and setting the same differently for different users(instances). 

Based on your statement I can conclude that CRON can be scheduled statically only :(  Having the dynamic scheduling feature would be very useful. I think quartz provide this feature via JobBuilder and Trigger components.

**Reply by sfmskywalker** [reply]
Ah I see. Actually, Elsa itself needs the ability to schedule jobs dynamically, and it uses Quartz under the covers. All the pieces are there.

If I actually understand your use case (do correct me if I'm wrong), you have one workflow definition that starts with a Cron activity, but you wish to dynamically trigger this workflow yourself and provide the actual cron expression as a parameter into the workflow.

If this is the case, then instead of using the Cron activity as the root activity, you should use an action-like activity as the root instead. For example, use `SetVariable` - this is convenient, because you can use it to set a variable called e.g. "MyCronExpression" to the value of the workflow input provided from your controller.

The second activity would be the Cron activity, configured using the cron expression stored in the MyCronExpression variable.

Now you will effectively be able to launch cron jobs manually from your controller and also control the cron expression itself.

I haven't tested any of this, but I will do so a little later and even write a small guide for the documentation website, because this is a very interesting use case :)

**Reply by SachinMehndiratta** [reply]
You understood this correctly, this is exactly the use case and the solution you are suggesting seems simple yet effective though I need to give this a run. I am also wondering about the use case to reschedule/stop the workflows(orchestration job). 

In simple words-

Let's say Workflow A is available in system starting with SetVariable and subsequent CRON activity

-  User A instantiate the workflow A, 8PM every night.
- User B instantiate the workflow A, 6 PM every Sunday.
- 1 week down the line, User A wants to change the schedule weekly i.e. to 8PM every Tuesday . For this step I need to identify the workflow instance created for User A in step 1 and the possibility to reschedule OR may be to stop the existing and invoking new instance
- Now User B wants to stop the job, i.e. capability to identify the instance created in step 2 and stopping the same

Once I am able to execute the above use case then I'll repeat the same but creating the Workflow A via dashboard instead of code.

I am relatively new to Elsa, started late last week only. Will go through the existing documentation.

Also, is there any documentation on arch of Elsa available? I wanted to look into source code and some initial directions would save hell lot of time.

**Reply by sfmskywalker** [reply]
Hi @SachinMehndiratta , I realize I never replied to you or wrote that guide I mentioned. Do you still need any of this?

To answer some of your other questions:

1. Once you have a workflow instance running, you can access it by ID using `IWorkflowInstanceStore` and manipulate the CRON activity's state manually (using code) and then save it back. This should update the Quartz jobs with the new CRON expression.
2. To stop a job, you could simply put the workflow instance into the Finished state and save it back to the DB. Alternatively, you could setup the workflow as a state machine that listens for signals. A signal could be "Pause", transitioning the workflow into a Paused state.
3. There's no documentation on Elsa's architecture at the moment. Almost everything revolves around the core `WorkflowRunner` class that executes Workflow Instances and their associated Workflow Blueprint. Everything else is about being able to identity what workflows to run based on some trigger, exposing APIs to the dashboard, persistence etc. I am of course working on increasing documentation coverage.
